src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,114 @@
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final int ONE = 1;
+    private static final int ZERO = 0;
+    private T[] values;
+    private int size;
+
+    public ArrayList() {
+        values = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        if (size == values.length) {
+            grow();
+        }
+        values[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
+        checkIndexForAdd(index);
+        if (size == values.length) {
+            grow();
+        }
+        System.arraycopy(values, index,
+                values, index + ONE,
+                size - index);
+        values[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        T[] a = (T[]) new Object[list.size()];
+        for (int i = 0; i < list.size(); i++) {
+            a[i] = list.get(i);
+        }
+        int varLength = a.length;
+        if (size == values.length) {
+            grow();
+        }
+        System.arraycopy(a,ZERO,values,size, varLength);
+        size += varLength;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return values[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndex(index);
+        values[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        final T removedElement = values[index];
+        System.arraycopy(values, index + ONE,
+                values, index,
+                values.length - index - ONE);
+        values[values.length - ONE] = null;
+        size--;
+        return removedElement;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        for (int i = 0; i < size; i++) {
+            if (Objects.equals(values[i], element)) {
+                return remove(i);
+            }
+        }
+        throw new NoSuchElementException("Element not found in ArrayList");
+    }
+
+    public void grow() {
+        T[] temp = (T[]) new Object[values.length];
+        System.arraycopy(values, ZERO, temp,ZERO, size);
+        values = (T[]) new Object[values.length << 1];
+        System.arraycopy(temp, ZERO, values,ZERO, size);
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == ZERO;
+    }
+
+    private void checkIndex(int index) {
+        if (index >= size || index < ZERO) {
+            throw new ArrayListIndexOutOfBoundsException("Can't find element by index");
+        }
+    }
+
+    private void checkIndexForAdd(int index) {
+        if (index > size || index < ZERO) {
+            throw new ArrayListIndexOutOfBoundsException("Index Out Of Bounds");
+        }
```

Comment:
these two methods do almost the same thing, it may be better to use one with the message ArrayListIndexOutOfBoundsException("Incorrect index: " + index)